pkg/jit/jitdump.go

```diff
@@ -0,0 +1,460 @@
+	CodeSize    uint64    // size in bytes of the jitted code
+	CodeIndex   uint64    // index referring to the JRCodeLoad CodeIndex record of when the function was initially jitted
+}
+
+// DebugEntry reprensents an entry from a JITCodeDebugInfo record.
+type DebugEntry struct {
+	Addr    uint64 // address of function for which the debug information is generated
+	Lineno  uint32 // source file line number (starting at 1)
+	Discrim uint32 // column discriminator, 0 is default
+	Name    string // source file name in ASCII
+}
+
+const debugEntryFixedSize int = 16 // size of DebugEntry fixed-sized fields
+
+// JRCodeDebugInfo represents a JITCodeDebugInfo record.
+type JRCodeDebugInfo struct {
+	Prefix   *JRPrefix     // the record header
+	CodeAddr uint64        // address of function for which the debug information is generated
+	NREntry  uint64        // number of debug entries for the function
+	Entries  []*DebugEntry // array of NREntry debug entries for the function
+}
+
+const jrCodeDebugInfoFixedSize int = jrPrefixSize + 16 // size of JRCodeDebugInfo fixed-sized fields
+
+// JRCodeUnwindingInfo represents a JITCodeUnwindingInfo record.
+type JRCodeUnwindingInfo struct {
+	Prefix         *JRPrefix // the record header
+	UnwindingSize  uint64    // the size in bytes of the unwinding data table at the end of the record
+	EHFrameHDRSize uint64    // the size in bytes of the DWARF EH Frame Header at the start of the unwinding data table at the end of the record
+	MappedSize     uint64    // the size of the unwinding data mapped in memory
+	UnwindingData  []byte    // an array of unwinding data, consisting of the EH Frame Header, followed by the actual EH Frame
+}
+
+const jrCodeUnwindingInfoFixedSize int = jrPrefixSize + 24 // size of JRCodeUnwindingInfo fixed-sized fields
+
+// JITDump represents the loaded jitdump.
+type JITDump struct {
+	Header        *JITHeader             // the jitdump file header
+	CodeLoads     []*JRCodeLoad          // JITCodeLoad records
+	CodeMoves     []*JRCodeMove          // JITCodeMove records
+	DebugInfo     []*JRCodeDebugInfo     // JITCodeDebugInfo records
+	UnwindingInfo []*JRCodeUnwindingInfo // JITCodeUnwindingInfo records
+}
+
+// jitDumpParser is used to parse a jitdump file.
+type jitDumpParser struct {
+	logger     log.Logger
+	buf        *bufio.Reader
+	endianness binary.ByteOrder
+}
+
+// newParser initializes a jitDumpParser.
+func newParser(logger log.Logger, rd io.Reader) (*jitDumpParser, error) {
+	p := &jitDumpParser{
+		logger: logger,
+	}
+
+	p.buf = bufio.NewReader(rd)
+
+	magic, err := p.buf.Peek(4)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read magic number from jitdump file: %w", err)
+	}
+
+	switch {
+	case bytes.Equal(magic, []byte{'J', 'i', 'T', 'D'}):
+		p.endianness = binary.BigEndian
+	case bytes.Equal(magic, []byte{'D', 'T', 'i', 'J'}):
+		p.endianness = binary.LittleEndian
+	default:
+		return nil, fmt.Errorf("%w: %#x", err, magic)
```

Comment:
Perhaps you forgot to use `ErrWrongJITDumpMagic` here?

Reply:
Indeed, excellent catch!